File: examples/ts/fixed-array-index/ArrayIndex.runar.ts
import { SmartContract, assert } from 'runar-lang';
import type { FixedArray } from 'runar-lang';

/**
 * ArrayIndex — the dynamic-index read of a `FixedArray`, on a stateless
 * contract, in all nine surface syntaxes.
 *
 * `Grid2x2` covers array literals and LITERAL-index access. It does not cover
 * `this.table[i]` where `i` is a runtime value, which is the path the
 * expand-fixed-arrays pass lowers into a chain of index comparisons — and the
 * path behind three of the four FixedArray defects this audit found.
 *
 * Stateless on purpose: it makes the fixture reachable by the differential
 * execution oracle (`conformance/witnesses/`), which drives spends through the
 * ANF interpreter and the @bsv/sdk ScriptVM and today has no stateful lane.
 */
export class ArrayIndex extends SmartContract {
  readonly table: FixedArray<bigint, 4> = [10n, 20n, 30n, 40n];

  constructor() {
    super();
  }

  public lookup(i: bigint, expected: bigint) {
    assert(this.table[i] === expected);
  }
}

File: examples/zig/fixed-array-index/ArrayIndex.runar.zig
const runar = @import("runar");

// ArrayIndex -- Zig port of
// examples/ts/fixed-array-index/ArrayIndex.runar.ts.
//
// Exercises [4]i64 together with a RUNTIME index read self.table[i].
pub const ArrayIndex = struct {
    pub const Contract = runar.SmartContract;

    table: [4]i64 = .{ 10, 20, 30, 40 },

    pub fn init() ArrayIndex {
        return .{};
    }

    pub fn lookup(self: *const ArrayIndex, i: i64, expected: i64) void {
        runar.assert(self.table[i] == expected);
    }
};
